concat-kdf/src/lib.rs

```diff
@@ -0,0 +1,110 @@
+//! An implementation of Concat KDF, the Concatenation Key Derivation Function.
+//!
+//! This function is described in the section 5.8.1 of [NIST SP 800-56A, Recommendation
+//! for Pair-Wise Key Establishment Schemes Using Discrete Logarithm Cryptography][1].
+//!
+//! # Usage
+//!
+//! The most common way to use Concat KDF is as follows: you generate a shared secret
+//! with other party (e.g. via Diffie-Hellman algorithm) and use key derivation function
+//! to derive a shared key.
+//!
+//! ```rust
+//! let mut key = [0u8; 32];
+//! concat_kdf::derive_key_into::<sha2::Sha256>(b"shared-secret", b"other-info", &mut key).unwrap();
+//! ```
+//!
+//! [1]: https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-56ar.pdf
+
+#![no_std]
+#![cfg_attr(docsrs, feature(doc_cfg))]
+
+use core::fmt;
+use digest::{generic_array::typenum::Unsigned, Digest, FixedOutputReset, Update};
+
+#[cfg(feature = "std")]
+extern crate std;
+
+/// Concat KDF errors.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum Error {
+    /// The length of the secret is zero.
+    NoSecret,
+    /// The length of the output is zero.
+    NoOutput,
+    /// The length of the output is too big.
+    CounterOverflow,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        let msg = match self {
+            Error::NoSecret => "Buffer for secret has zero length.",
+            Error::NoOutput => "Buffer for key has zero length.",
+            Error::CounterOverflow => "Requested key length is to big.",
+        };
+        f.write_str(msg)
+    }
+}
+
+#[cfg(feature = "std")]
```

Comment:
depending on what you need from std, having alloc available may be enough. e.g. if only vec or box is needed.
It can be handy for porting the lib if only alloc is needed and not the entire std. Depends on the needs though.

https://doc.rust-lang.org/std/alloc/index.html